app/src/main/java/com/example/mysimplecoindeck/ui/CoinsViewModel.kt

```diff
@@ -45,15 +53,57 @@ class CoinsViewModel @Inject constructor(
             repository.coinDetail(uuid)
                     .catch { exception ->
                         _uiCoinDetailState.value = CoinDetailUiState.Error(exception)
-                        Log.d("API-Test","error while call api: ${exception.message}")
+                        Log.d("GetCoinDetails","error while call api: ${exception.message}")
                     }
                     .collect{ coinDetail ->
                         _uiCoinDetailState.value = CoinDetailUiState.Success(coinDetail)
-                        Log.d("API-Test passed",coinDetail.body()?.status.toString())
+                        Log.d("GetCoinDetails passed",coinDetail.body()?.status.toString())
+                    }
+        }
+    }
+
+    fun getSuggestedCoins(query: String) {
+        viewModelScope.launch {
+            repository.getSearchSuggestions(query)
+                .catch { exception ->
+                    _uiSearchState.value = SearchUiState.Error(exception)
+                    Log.d("GetSuggestions","error while call api: ${exception.message}")
+                }
+                .collect{ response ->
+                    _uiSearchState.value = SearchUiState.Success(response)
+                    Log.d("GetSuggestions passed",response.body()?.status.toString())
+                }
+        }
+    }
+
+    fun getPortfolio() {
+        viewModelScope.launch {
+            repository.getPortfolio()
+                    .catch { exception ->
+                        _uiPortfolioState.value = PortfolioUiState.Error(exception)
+                        Log.d("Portfolio-DB","error while getting Portfolio: ${exception.message}")
+                    }
+                    .collect{ response ->
+                        _uiPortfolioState.value = PortfolioUiState.Success(response)
+                        Log.d("Portfolio-DB passed:", "Successfully got portfolio")
                     }
         }
     }
 
+    fun insertCoinToPortfolio(coinPortfolioEntity: CoinPortfolioEntity) {
```

Comment:
I think it would be better to move the `Entity` up and away from the view as much as possible -> pass the `Coin` here and convert it to the Entity here or even in the Repository. 

In general the view layer (fragment) should have almost no knowledge about the data layers, Entities or retrofit models should mapped to domain/view models as early as possible and vice-versa. Ideally only the Repository or ViewModel knows about both models, the view only knows the view models, the data layer only knows the entity model. 

Extension functions are quite useful for that, a `Mappers.kt` file can contain all mapping operations like: 

```kotlin
fun Coin.mapToEntity(): CoinPortfolioEntity {
    return CoinPortfolioEntity(uuid = this.uuid, ...)
}
fun CoinPortfolioEntity.mapToDomain(): Coin {
    return Coin(uuid = this.uuid, ...)
}
```

then the ViewModel or Repository can just call `coin.mapToEntity()` to get the matching entity model or `coinEntity.mapToDomain()` for the domain/view model.

This approach also has the benefit, that changing the model (adding/removing a parameter) is quite easy as you only need to change the mapper once and not everywhere.